### ISSUE TITLE
Eureka! Code Profiling

### ISSUE THREAD

**Issue by jbrande** [kevin218/Eureka#308]
### What happened?

We need to profile Eureka's performance in order to assess major bottlenecks and hopefully speed things up. This is especially important for Stage 3 and Stage 5. I think I can do this with Python's cProfile utility, so that's where I'll be starting.

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

**Comment by jbrande** [comment]
Looking at S5 emcee runs now. Most of the time spent is spent in likelihood.py, but a lot of this actually looks to be built-in python/numpy stuff we probably don't have much control over. However, about 50 seconds of the 135 seconds that we're in likelihood.py is spent in Model.py either evaluating or updating the models. In update(), a good chunk of this is spent reading EPFs, and parsing the coefficients.

![s5_profile](https://user-images.githubusercontent.com/5041139/168348058-ab3cfbbc-ce1f-42a0-9512-0c5635bbfa91.jpg)

In eval(), the biggest time sink seems to be evaluating the transit and systematics models. The Batman transit model takes more of this time than the polynomial systematics model, so perhaps when starry gets implemented, we might be able to speed up the transit lightcurves a bit.

**Comment by taylorbell57** [comment]
This is super interesting Yoni, thanks so much for looking into this! I think improving the update function will help a lot and should be pretty doable. I've been thinking for a while that we're really not doing things quickly in that function, but it's nice to see just how big an effect this has.

For S3 I definitely agree that a tonne of time is spent on plotting. The plot resolution size might have some effect on this, but I wonder if that's the main issue or if there's something else we can do. I guess it is the savefig lines that are the slowest though, so maybe resolution really is the bigger concern. I wonder if PDF vs PNG significantly changes things and if there's any other optimizations like that we can do (like different backends as you'd mentioned).

**Comment by taylorbell57** [comment]
I quite strongly suspect that a significant part of issue #223 is caused by the inefficient update functions in Stage 5, but it's possible dynesty just scales very poorly with the number of data and/or fitted parameters

**Comment by jbrande** [comment]
Yeah, getting into the dynesty profile it's similar to the emcee profile, except that the nested sampling prior transforms also take a good bit of time. 
![s5_dynesty_profile](https://user-images.githubusercontent.com/5041139/168491451-1b621acc-c4b2-42eb-af8a-56cd7954278b.jpg)

Reading EPFs, setting attributes, transit and systematics models all still seem to be where we have room to improve.

**Comment by jbrande** [comment]
Just a short update, the proportion of time spent in each operation in S5 is basically the same between instruments for a given set of transit/systematics parameters. It's not super surprising but it's nice to confirm this. I've tested this with NIRCam and NIRSpec, and now I'll be branching off into the different eclipse/phase curve/GP possibilities with MIRI.

**Comment by taylorbell57** [comment]
I used this profiling information to speed up S5 with PR #434 and looking at the S3 profiling information I don't think there is anything overly worth trying to improve there. @jbrande do you think we should keep this issue open, or do you agree that we can close it now?

**Comment by kevin218** [comment]
We should use Scalene, an AI based profiling tool that is available on GitHub.
